src/consensus/proposer.rs

```diff
@@ -58,10 +58,56 @@ pub trait Proposer {
     fn get_proposed_value(&self, shard_hash: &ShardHash) -> Option<FullProposal>;
 }
 
+pub struct ProposedValues {
+    values_by_height: BTreeMap<(Height, Round), ShardHash>,
+    values: BTreeMap<ShardHash, FullProposal>,
+}
+
+impl ProposedValues {
```

Comment:
Worth adding a test for this. 

We should also add tests for the proposers themselves, but that might require a bunch of setup work. So, ok to defer for now